salt/orch/update.sls

```diff
@@ -0,0 +1,117 @@
+# Generic Updates
+update_pillar:
+  salt.function:
+    - tgt: '*'
+    - name: saltutil.refresh_pillar
+
+# TODO: When we update to 2016.11.4 uncomment this
+# update_grains:
+#   salt.function:
+#     - tgt: '*'
+#     - name: saltutil.refresh_grains
+
+update_mine:
+  salt.function:
+    - tgt: '*'
+    - name: mine.update
+    - require:
+       - salt: update_pillar
+
+update_modules:
+  salt.function:
+    - name: saltutil.sync_modules
+    - tgt: '*'
+    - kwarg:
+        refresh: True
+
+# Get list of masters needing reboot
+{%- set masters = salt.saltutil.runner('mine.get', tgt='G@roles:kube-master and G@tx_update_reboot_needed:true', fun='network.ip_addrs', tgt_type='compound') %}
+
+{%- for master_id in masters.keys() %}
+
+{{ master_id }}-clean-shutdown:
+  salt.state:
+    - tgt: {{ master_id }}
+    - sls:
+      - kubernetes-master.stop
+      - flannel.stop
+      - etcd.stop
+
+# Reboot the node
+{{ master_id }}-reboot:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: cmd.run
+    - arg:
+      - rebootmgrctl reboot now
+
+# Wait for it to start again
+{{ master_id }}-wait-for-start:
+  salt.wait_for_event:
+    - name: salt/minion/*/start
+    - id_list:
+      - {{ master_id }}
+
+# Start services
+{{ master_id }}-start-services:
+  salt.state:
+    - tgt: {{ master_id }}
+    - highstate: True
+
+{{ master_id }}-update-grain:
+  salt.function:
+    - tgt: {{ master_id }}
+    - name: grains.setval
+    - arg:
+      - tx_update_reboot_needed
+      - false
+    - require:
+      - salt: {{ master_id }}-start-services
+
+{% endfor %}
+
+{%- set workers = salt.saltutil.runner('mine.get', tgt='G@roles:kube-minion and G@tx_update_reboot_needed:true', fun='network.ip_addrs', tgt_type='compound') %}
+{%- for worker_id, ip in workers.items() %}
+
+# Call the node clean shutdown script
+{{ worker_id }}-clean-shutdown:
+  salt.state:
+    - tgt: {{ worker_id }}
+    - sls:
+      - kubernetes-minion.stop
+      - docker.stop
+      - flannel.stop
+      - etcd-proxy.stop
+
+# Reboot the node
+{{ worker_id }}-reboot:
+  salt.function:
+    - tgt: {{ worker_id }}
+    - name: cmd.run
+    - arg:
+      - rebootmgrctl reboot now
+
+# Wait for it to start again
+{{ worker_id }}-wait-for-start:
+  salt.wait_for_event:
+    - name: salt/minion/*/start
+    - id_list:
+      - {{ worker_id }}
+
+# Start services
+{{ worker_id }}-start-services:
+  salt.state:
+    - tgt: {{ worker_id }}
+    - highstate: True
+
+{{ worker_id }}-update-grain:
+  salt.function:
+    - tgt: {{ worker_id }}
+    - name: grains.setval
+    - arg:
+      - tx_update_reboot_needed
+      - false
+    - require:
+      - salt: {{ worker_id }}-start-services
```

Comment:
Maybe they are not strictly necessary but I would include some `require` for forcing some order... otherwise, you could add a compact form:

```SaltStack
{{ worker_id }}-bootup:
  salt.wait_for_event:
    - name: salt/minion/*/start
    - id_list:
      - {{ worker_id }}
  salt.state:
    - tgt: {{ worker_id }}
    - highstate: True
  salt.function:
    - tgt: {{ worker_id }}
    - name: grains.setval
    - arg:
      - tx_update_reboot_needed
      - false
```